File: AlgorithmReview/Programmers/Lv2/괄호 회전하기.swift
//
//  괄호 회전하기.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/03/29.
//

import Foundation

func solution(_ s:String) -> Int {

    let bracket = ["()", "[]", "{}"]
    let openBracket = bracket.map { $0.first! }

    func isCorrect(_ rotation: String) -> Bool{
        var stack = [Character]()
        for c in rotation{
            if(openBracket.contains(c)){
                stack.append(c); continue
            }
            let last = stack.last ?? " "
            if(bracket.contains("\(last)\(c)")){
                stack.removeLast()
            }else{
                return false
            }
        }
        return stack.isEmpty ? true : false
    }

    var string = Array(s)
    var ans = 0
    for i in 0..<s.count{
        let rotation = string[i..<s.count] + string[0..<i]
        if(isCorrect(String(rotation))){
            ans += 1
        }
    }
    return ans
}

File: Python/괄호 회전하기.py
def solution(s):
    
    s = list(s)
    bracket = ["()","[]","{}"]
    result = 0
    
    for i in range(0, len(s)):
        
        stack = [s[i]]
        
        j = 0 if i+1 == len(s) else i+1
        
        while j != i:
            
            if len(stack) != 0 and (stack[-1]+s[j] in bracket):
                stack.pop()
            else:
                stack.append(s[j])
                
            j = 0 if j+1 == len(s) else j+1
            
        if len(stack) == 0:
            result += 1
    
    return result
